plugins/obs-webrtc/whep-source.cpp

```diff
@@ -0,0 +1,511 @@
+			  std::back_inserter(msg));
+		this->sps_and_pps = std::vector<std::byte>{};
+	}
+
+	pkt->data = reinterpret_cast<uint8_t *>(msg.data());
+	pkt->size = static_cast<int>(msg.size());
+
+	auto ret = avcodec_send_packet(this->video_av_codec_context.get(), pkt);
+	if (ret < 0) {
+		return;
+	}
+
+	AVFrame *av_frame = this->av_frame.get();
+
+	while (true) {
+		ret = avcodec_receive_frame(this->video_av_codec_context.get(),
+					    av_frame);
+		if (ret < 0) {
+			if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+				break;
+			}
+			return;
+		}
+
+		struct obs_source_frame frame = {};
+
+		frame.format = VIDEO_FORMAT_I420;
+		frame.width = av_frame->width;
+		frame.height = av_frame->height;
+		frame.timestamp = pts;
+		frame.max_luminance = 0;
+		frame.trc = VIDEO_TRC_DEFAULT;
+
+		video_format_get_parameters_for_format(
+			VIDEO_CS_DEFAULT, VIDEO_RANGE_DEFAULT, frame.format,
+			frame.color_matrix, frame.color_range_min,
+			frame.color_range_max);
+
+		for (size_t i = 0; i < MAX_AV_PLANES; i++) {
+			frame.data[i] = av_frame->data[i];
+			frame.linesize[i] = abs(av_frame->linesize[i]);
+		}
+
+		obs_source_output_video(this->source, &frame);
+		last_frame = std::chrono::system_clock::now();
+	}
+}
+
+void WHEPSource::SetupPeerConnection()
+{
+	peer_connection = std::make_shared<rtc::PeerConnection>();
+	peer_connection->onStateChange([this](rtc::PeerConnection::State state) {
+		switch (state) {
+		case rtc::PeerConnection::State::New:
+			do_log(LOG_INFO, "PeerConnection state is now: New");
+			break;
+		case rtc::PeerConnection::State::Connecting:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Connecting");
+			break;
+		case rtc::PeerConnection::State::Connected:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Connected");
+			break;
+		case rtc::PeerConnection::State::Disconnected:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Disconnected");
+			break;
+		case rtc::PeerConnection::State::Failed:
+			do_log(LOG_INFO, "PeerConnection state is now: Failed");
+			break;
+		case rtc::PeerConnection::State::Closed:
+			do_log(LOG_INFO, "PeerConnection state is now: Closed");
+			break;
+		}
+	});
+
+	rtc::Description::Audio audioMedia(
+		"0", rtc::Description::Direction::RecvOnly);
+	audioMedia.addOpusCodec(111);
+	audio_track = peer_connection->addTrack(audioMedia);
+
+	auto audio_depacketizer = std::make_shared<rtc::OpusRtpDepacketizer>();
+	auto audio_session = std::make_shared<rtc::RtcpReceivingSession>();
+	audio_session->addToChain(audio_depacketizer);
+	audio_track->setMediaHandler(audio_depacketizer);
+	audio_track->onFrame([&](rtc::binary msg, rtc::FrameInfo frame_info) {
+		this->OnFrameAudio(msg, frame_info);
+	});
+
+	rtc::Description::Video videoMedia(
+		"1", rtc::Description::Direction::RecvOnly);
+	videoMedia.addH264Codec(96);
+	video_track = peer_connection->addTrack(videoMedia);
+
+	auto video_depacketizer = std::make_shared<rtc::H264RtpDepacketizer>();
```

Comment:
Let me share an issue related to `H264RtpDepacketizer` and a possible solution to it.

It turned out, in the following scenario, `H264RtpDepacketizer` could crash during processing incoming packets:

1. Connect OBS WHEP client to a SFU server.
2. Publish a stream from Chrome (to the above waiting WHEP client via SFU). 
3. The WHEP client receives video packets having empty payload.
   - I don't know why but Chrome seems to sometimes send such empty payload packets
5. `H264RtpDepacketizer` crashes, and the following log message is emitted by `libdatachannel`
   - `rtc::impl::DtlsTransport::doRecv@598: DTLS recv: MbedTLS error: SSL - Internal error (eg, unexpected failure in lower-level module)` 

I'm not sure this is the best solution, but I created the following patch to filter out such empty payload video packets before `H264RtpDepacketizer` processes those packets. 

```diff
diff --git a/plugins/obs-webrtc/whep-source.cpp b/plugins/obs-webrtc/whep-source.cpp
index 2f6d5ebf2..026af0a0b 100644
--- a/plugins/obs-webrtc/whep-source.cpp
+++ b/plugins/obs-webrtc/whep-source.cpp
@@ -348,6 +348,38 @@ void WHEPSource::OnFrameVideo(rtc::binary msg, rtc::FrameInfo frame_info)
        }
 }

+class RTC_CPP_EXPORT VideoDepacketizer : public rtc::H264RtpDepacketizer {
+public:
+  VideoDepacketizer() = default;
+  virtual ~VideoDepacketizer() = default;
+
+  void incoming(rtc::message_vector &messages, const rtc::message_callback &send) override{
+    messages.erase(std::remove_if(messages.begin(), messages.end(),
+                                  [&](rtc::message_ptr message) {
+                                    if (message->type == rtc::Message::Control) {
+                                      return false;
+                                    }
+
+                                    if (message->size() < sizeof(rtc::RtpHeader)) {
+                                      return false;
+                                    }
+
+                                    auto pkt = message;
+                                    auto pktParsed = reinterpret_cast<const rtc::RtpHeader *>(pkt->data());
+                                    auto headerSize =
+                                      sizeof(rtc::RtpHeader) + pktParsed->csrcCount() + pktParsed->getExtensionHeaderSize();
+                                    if (message->size() <= headerSize) {
+                                      // Discard empty payload packet to prevent error in the following `rtc::H264RtpDepacketizer::incoming()` call.
+                                      return true;
+                                    }
+
+                                    return false;
+                                  }),
+                   messages.end());
+    rtc::H264RtpDepacketizer::incoming(messages, send);
+  }
+};
+
 void WHEPSource::SetupPeerConnection()
 {
        peer_connection = std::make_shared<rtc::PeerConnection>();
@@ -395,7 +427,7 @@ void WHEPSource::SetupPeerConnection()
        videoMedia.addH264Codec(96);
        video_track = peer_connection->addTrack(videoMedia);

-       auto video_depacketizer = std::make_shared<rtc::H264RtpDepacketizer>();
+       auto video_depacketizer = std::make_shared<VideoDepacketizer>();
        auto video_session = std::make_shared<rtc::RtcpReceivingSession>();
        video_session->addToChain(video_depacketizer);
        video_track->setMediaHandler(video_depacketizer);
```

Reply:
@sile Nice catch! Do you want me to submit bug fix to libdatachannel, or do you want to make PR?

I would love to have you involved/make the PR! We have so much work left, so I could use the help :)

If you don't have the time I can do it though!

Reply:
Thank you!
I can make the PR for `libdatachannel` 💪

Reply:
Created: https://github.com/paullouisageneau/libdatachannel/pull/1140